### DISCUSSION TITLE
Integration with Zod and/or other data validation libraries

### DISCUSSION THREAD

**Discussion by Emilcrafter** [tywalch/electrodb#243] [Ideas]
I am in the ideation stages of a project to build an open-source CQRS Event Sourcing framework for DynamoDB. Parts of the library will require similar functionality to what ElectroDB offers. I had envisioned building something very similar, but with a neater process for schemas. For example, using Zod. From what I've seen, there shouldn't really be any compatibility issues between ElectroDB and Zod. Even "default values" are possible to implement. If the maintainers of this project are interested and able to provide some support, this is something I would be interested in collaborating on (instead of building functionality that already exists). So please enlighten me on whether there is any interest in creating first-class Zod (Or another data validation library if there are better options) integration in ElectroDB).

**Comment by zirkelc** [comment]
I'm looking for exactly the same functionality. Defining the schema with Zod seems like the perfect fit.

**Comment by tywalch** [comment]
This has been asked for before, but I am still a bit unclear what that would look like. Is your thought that an Entity would accept a Zod schema instead of a Schema?

**Reply by zirkelc** [reply]
I personally use [tRPC](https://github.com/trpc/trpc) for the APIs and it uses Zod for defining and validating input and output types. The great thing about [Zod](https://github.com/colinhacks/zod) is that you can define a Zod schema object like `User` once and use it to validate any input on the frontend **and** backend. If you need subtypes like `UserInput`, you can simply derive those types from the main object by picking or omitting certain fields, as you can in TypeScript. And you get static TypeScript types that are automatically derived based on the schema via `z.infer<typeof schema>`. 

Here's an example of the Book schema from Electro converted to Zod:
https://stackblitz.com/edit/typescript-gytttd?file=index.ts&view=editor

```ts
const zodBook = z.object({
  storeId: z.string(),
  bookId: z.string().default(() => randomUUID()),
  price: z.number().positive(),
  title: z.string().refine((val) => val.length <= 255, {
    message: "Title can't be more than 255 characters",
  }),
  author: z.string(),
  condition: z.enum(['EXCELLENT', 'GOOD', 'FAIR', 'POOR']),
  genre: z.set(z.string()),
  published: z.number().optional(),
});

type Book = z.infer<typeof zodBook>;
```

I really think that the ability to use a Zod schema for ElectroDB is a complete game changer. Theoretically it would allow to use the same schema for types, validation and parsing on the frontend, backend and database layer.

**Reply by Emilcrafter** [reply]
This is exactly what I meant. It seems fully compatible, no?

**Reply by tywalch** [reply]
Sorry, when I said "what it would look like" I suppose I meant how the schema would be used beyond defining the schema: validations and typing performed on "update" and "put" are different. For example, a "required=true" property cannot be deleted, and a "readOnly=true" property can be used in a "put" but not "update"

**Reply by tywalch** [reply]
When I think about the user requiring zod I think the intent must be to use the zod type outside of ElectroDB? That's the case I'm referring to, the type you'd use to define schema won't always match up to what you can use with different operations. It may work to validate a "put", but what about "update"?

**Reply by Emilcrafter** [reply]
Those are some good points. I see how that would pose a challenge. The first thing that comes to mind is adding a prefix to field names, like "_field" for readonly fields and "$field" for required=true. This could possibly be implemented using Zod's transform method, which can validate an input against a type and then add some additional data to it or change it in some other way. This needs further investigation into how it could be done in practice. I certainly think there exists some good way to do this without creating unneccessarily messy code.

**Reply by tywalch** [reply]
What problem is accepting a zod schema trying to solve? My assumption is that it has something to do with a desire to use zod elsewhere like validation middleware? Assuming you could pass a zod object to electro, what would an app like that look like? Would you be able to throw together a gist?
